ts: Convert `web/src/billing` module to TypeScript.

Lalit3716 opened the pull request:
Converted all files inside `web/src/billing` to TypeScript.

<!-- Describe your pull request here.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
@mateuszmandera can you review this one? Feel free to take some time learning TypeScript if you don't know it well.

andersk commented on web/src/billing/helpers.ts:
```diff
@@ -39,11 +57,11 @@ export function create_ajax_request(
         url,
         data,
-        success(response) {
+        success(response: T) {
```
This `response` parameter comes from a value of type `any`, so its validity as type `T` is completely unchecked. We should write `response: unknown`, and the caller should use [Zod](https://zod.dev/) to safely parse and validate it as the desired type.

Lalit3716 replied:
@andersk Thanks for the review I have integrated Zod but how do I fix failing tests now should I make every field nullable in my Zod schema? I would like your opinions on this!

PIG208 commented on web/src/page_params.ts:
```diff
@@ -37,6 +37,10 @@ export const page_params: {
+    annual_price: number;
+    monthly_price: number;
+    percent_off: number;
+    seat_count: number;
```
The `page_params` for the billing page is different from the homepage one, albeit both being `page_params`. Maybe we can create a separate `page_params` for billing only.
https://github.com/zulip/zulip/blob/9e4529c48bee7da41b32888d5f055f7c070b9078/corporate/views/upgrade.py#L274-L280

Lalit3716 replied:
I don't think I fully understand what's going on here 😅
I mean doesen't this context is being used for rendering the `corporate/upgrade.html` template only?
Can you elaborate a little that why and how were we using same `page_params` if they were different from the start!

PIG208 replied:
Right. `page_params.ts` is also used in other places where these properties are not available. So adding `seat_count` and etc to `page_params` would not be correct in general.

PIG208 replied:
We can avoid importing or changing `page_params.ts`, and in a prep commit, access `$("#page-params").remove().data("params")` directly, and extract the needed fields to be passed to `initialize` in `upgrade.js`.

Lalit3716 replied:
Oh, okay that makes sense to me, but we also need `page_params.seat_count` data in `helper.js` also will it make more sense to make a new separate `page_params` for billing?

PIG208 replied:
Yeah, that sounds reasonable for now. Eventually, we will get rid of `page_params`.

Lalit3716 replied:
Yeah that sounds good I will try to make a seperate `page_params` in billing directory for now!

Lalit3716 replied:
@PIG208 So I was trying to access `$("#page_params")` in a newly created `page_params.ts` in billing directory but of course, I couldn't access it because for now, we are removing `#page_params` element in the main `page_params.ts`, any suggestions, should I change main `page_params.ts` to not remove the `#page_params` element from the DOM?

PIG208 replied:
I think you shouldn't need to change the main `page_params.ts`. We only remove `#page_params` once upon initialization. You are just replacing the main `page_params.ts` with a local version for billing. It should function exactly the same way.

Lalit3716 replied:
Alright! so made a new commit introducing new `page_params` as well as adjusted the testing code for it!

PIG208 reviewed (commented):
I left some comments on how to fix the failing test cases as well as some other issues. @Lalit3716  Great work!

Lalit3716 commented on web/src/billing/event_status.ts:
```diff
@@ -1,10 +1,28 @@
-function update_status_and_redirect(status_message, redirect_to) {
+const stripe_response_schema = z.object({
+    session: z.object({
+        id: z.string(),
```
Sorry! my fault I don't know how that id slipped in there 😅

Lalit3716 commented on web/src/billing/event_status.ts:
```diff
@@ -125,30 +169,33 @@ export async function stripe_payment_intent_status_check(stripe_payment_intent_i
-        setTimeout(start_status_polling, 5000);
+        // If the request fails, we just retry after 5 seconds.
+        await new Promise((resolve) => setTimeout(resolve, 5000));
+        await start_status_polling();
```
This change was mainly done due to eslint complaining about the rule "no-misused-promises" though it can also be avoided by just placing a void keyword before `start_status_polling` in `setTimeout` like this:
`setTimeout(() => void start_status_polling(), 5000);`
I think this might bit be more efficient than doing it like above change?

PIG208 replied:
Yeah, this looks cleaner.

Lalit3716 commented on web/src/billing/helpers.ts:
```diff
@@ -3,13 +3,31 @@ import $ from "jquery";
 import * as loading from "../loading";
 import {page_params} from "../page_params";
 
+type FormDataObject = {
```
Agreed!

Lalit3716 commented on web/src/billing/event_status.ts:
```diff
@@ -1,10 +1,28 @@
+        type: z.string(),
+        stripe_payment_intent_id: z.string(),
+        status: z.string(),
+        event_handler: z.object({
```
Thanks for pointing this file out!

Lalit3716 commented:
Thanks for the thorough review @PIG208! I fixed some of the issues you mentioned but I have some doubts on some of them and hence left some comments please take a look 🙂!

PIG208 commented on package.json:
```diff
@@ -87,7 +87,8 @@
     "webpack-cli": "^5.0.1",
-    "winchan": "^0.2.1"
+    "winchan": "^0.2.1",
+    "zod": "^3.20.6"
```
We also want to bump [PROVISION_VERSION](https://zulip.readthedocs.io/en/latest/subsystems/dependencies.html#provision-version) by a minor level since we add a dependency here.

PIG208 commented:
The new commit should be a prep change before the TypeScript conversion, so we can avoid having an intermediate state where the types are inaccurate.

Lalit3716 commented:
@PIG208 reordered the commits!

PIG208 commented on version.py:
```diff
@@ -48,4 +48,4 @@
 #   minor version bump suffices.
 
-PROVISION_VERSION = (227, 0)
+PROVISION_VERSION = (228, 0)
```
A minor bump should suffice, i.e. `(227, 1)`.

Lalit3716 replied:
Also It looks like @andersk already added Zod as a dependency in this [commit](https://github.com/zulip/zulip/commit/38539c8fed521c6e3f914a0b4709b00e3157db28)!

PIG208 commented on web/src/billing/upgrade.ts:
```diff
@@ -0,0 +1,104 @@
+    event: string;
+    education: string;
+    education_nonprofit: string;
+};
```
Instead of copying over these types on both `upgrade.ts` and `helpers.ts`, it is better to just export these types from `helpers.ts` and just use them here.

PIG208 commented on web/tests/event_status.test.js:
```diff
@@ -37,6 +37,7 @@ run_test("check_status", async ({override}) => {
             session: {
                 status: "created",
+                type: "upgrade_from_billing_page",
             },
```
Just want to confirm: are there any circumstances in which `type` does not get returned?

Lalit3716 replied:
I have checked the view for this request, and yes, there isn't any case where a `type` is not returned from the request.

Lalit3716 replied:
Except of course when we pass `stripe_payment_intent_id` in request params instead of `stripe_session_id` in which case it returns a `payment_intent` instead of a `session`.

PIG208 commented on web/src/billing/upgrade.ts:
```diff
@@ -0,0 +1,104 @@
+
+type StripeSessionResponse = {
+    stripe_session_url: string;
+};
```
We should move its corresponding schema into `helpers.ts` and use the inferred type of that schema, instead of manually rewriting `StripeSessionResponse`. Same thing in `event_status.ts`.
```
const response_schema = z.object({
    stripe_session_url: z.string(),
});
```

PIG208 commented on web/src/billing/event_status.ts:
```diff
@@ -82,41 +105,70 @@ export function initialize_retry_with_another_card_link_click_handler() {
-                window.location.replace(response.stripe_session_url);
+                const response_schema = z.object({
+                    stripe_session_url: z.string(),
+                });
```
This particular schema can be moved into `helpers.ts`, because it is shared between `event_status.ts` and `upgrade.ts`.

Lalit3716 commented:
Thanks for the review! @PIG208 , resolved all suggestions!

Lalit3716 commented:
Hi, @PIG208 I was revisiting this PR and found some issues that we needed to address by making a separate `page_params` for billing:-
1) We needed to include `./src/billing/page_params` in `webpack.assets.json` before anything so that the `#page_params` does not get removed from the DOM by the main `page_params` by the time we imported new `page_params`.
2) Current `webpack.assets.json`:
      ```json
          "upgrade": [
              "./src/billing/page_params",
              "./src/bundles/portico",
              "./src/portico/tippyjs",
              "./src/portico/landing-page",
              "./styles/portico/landing_page.css",
              "./src/billing/helpers",
              "./src/billing/upgrade",
              "./src/templates",
              "./src/loading",
              "./styles/portico/billing.css"
          ],
      ```
    Now if we remove `#page_params` in `billing/page_params` then we will not have them available for example in `src/bundles/portico` and hence we don't want to remove the `#page_params` from the DOM in the new billing `page_params` because they are needed by other modules later down the line.

I have updated the PR addressing these issues and hence this is ready for another review from you, Thanks 🙂!

PIG208 reviewed (commented):
Posted some comments. Overall this looks good!

Lalit3716 commented on web/src/page_params.ts:
```diff
@@ -57,6 +57,10 @@ export const page_params: {
+    annual_price: number;
+    monthly_price: number;
+    percent_off: number;
+    seat_count: number;
```
Oh yeah right I forgot to remove them, fixed it!

Lalit3716 commented on web/tests/billing.test.js:
```diff
@@ -60,6 +65,7 @@ run_test("card_update", ({override}) => {
     const update_card_click_handler = $("#update-card-button").get_on_handler("click");
     override(helpers, "create_ajax_request", card_change_ajax);
     update_card_click_handler({preventDefault() {}});
+    console.log("card_update");
```
Yeah idk how that slipped in 😅

PIG208 commented:
@zulipbot add "integration review"

timabbott commented:
This is waiting for someone who knows the billing subsystem better than me to at least test it; probably @mateuszmandera?

mateuszmandera commented:
> This is waiting for someone who knows the billing subsystem better than me to at least test it; probably @mateuszmandera?

Yup, will give it some testing and review

Lalit3716 commented:
@zulipbot add "area: typescript migration"

timabbott commented:
@mateuszmandera did you complete the testing/review for this?

mateuszmandera commented:
@timabbott Yes, it works for me in testing and the code changes make sense to me, to the extent of my limited TypeScript understanding.

shameondev commented on web/src/billing/event_status.ts:
```diff
@@ -1,10 +1,31 @@
 import * as helpers from "./helpers";
 
-function update_status_and_redirect(status_message, redirect_to) {
+const stripe_response_schema = z.object({
```
Are these changes directly related to the TypeScript migration? If not, I think it would be better to make a separate commit to introduce these changes.

Lalit3716 replied:
Also, these changes aren't exactly refactoring changes as well they only were made to support typescript migration only.

shameondev commented:
As far as I can see, @mateuszmandera has completed the review with a positive result. Therefore, I am adding the "integration review" label.

timabbott commented:
Merged, thanks everyone!
